From 1fc42297a63e1a867487595988289e11b96ab636 Mon Sep 17 00:00:00 2001
From: Brahma Reddy Battula <brahma@apache.org>
Date: Tue, 20 Jun 2017 20:44:31 +0800
Subject: [PATCH 2556/2748] HADOOP-14533. Size of args cannot be less than
 zero in TraceAdmin#run as its linkedlist.
 Contributed by Weisen Han.

(cherry picked from commit 2e9daa2e27536484846d9c5f020c6afaf27932ce)
(cherry picked from commit a526c3b0bfe42e340a21ea06d36f52a1c6231abd)

Change-Id: Ic7dae02de24482956b282dbcbc0106e767ed1461
---
 .../java/org/apache/hadoop/tracing/TraceAdmin.java |    2 +-
 .../org/apache/hadoop/tracing/TestTraceAdmin.java  |    8 ++++++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/TraceAdmin.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/TraceAdmin.java
index 4ae5aed..512c7ac 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/TraceAdmin.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/TraceAdmin.java
@@ -159,7 +159,7 @@ public int run(String argv[]) throws Exception {
       System.err.println("You must specify a host with -host.");
       return 1;
     }
-    if (args.size() < 0) {
+    if (args.isEmpty()) {
       System.err.println("You must specify an operation.");
       return 1;
     }
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java
index 198dafb..e356423 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java
@@ -59,6 +59,14 @@ private String getHostPortForNN(MiniDFSCluster cluster) {
   }
 
   @Test
+  public void testNoOperator() throws Exception {
+    TraceAdmin trace = new TraceAdmin();
+    trace.setConf(new Configuration());
+    Assert.assertEquals("ret:1, You must specify an operation." + NEWLINE,
+        runTraceCommand(trace, "-host", "127.0.0.1:12346"));
+  }
+
+  @Test
   public void testCreateAndDestroySpanReceiver() throws Exception {
     Configuration conf = new Configuration();
     conf = new Configuration();
-- 
1.7.9.5

